pandas/tests/indexes/period/test_period_range.py

```diff
@@ -0,0 +1,54 @@
+import pytest
+import pandas.util.testing as tm
+from pandas import date_range, period_range, PeriodIndex
+
+
+class TestPeriodRange(object):
+
+    @pytest.mark.parametrize('freq', ['D', 'W', 'M', 'Q', 'A'])
+    def test_construction(self, freq):
+        # non-empty
+        expected = date_range(start='2017-01-01', periods=5,
+                              freq=freq, name='foo').to_period()
+        start, end = str(expected[0]), str(expected[-1])
+
+        result = period_range(start=start, end=end, freq=freq, name='foo')
+        tm.assert_index_equal(result, expected)
+
+        result = period_range(start=start, periods=5, freq=freq, name='foo')
+        tm.assert_index_equal(result, expected)
+
+        result = period_range(end=end, periods=5, freq=freq, name='foo')
+        tm.assert_index_equal(result, expected)
+
+        # empty
+        expected = PeriodIndex([], freq=freq, name='foo')
+
+        result = period_range(start=start, periods=0, freq=freq, name='foo')
+        tm.assert_index_equal(result, expected)
+
+        result = period_range(end=end, periods=0, freq=freq, name='foo')
+        tm.assert_index_equal(result, expected)
+
+        result = period_range(start=end, end=start, freq=freq, name='foo')
+        tm.assert_index_equal(result, expected)
+
+    def test_errors(self):
+        # not enough params
+        msg = ('Of the three parameters, start, end, and periods, '
+               'exactly two must be specified')
+        with tm.assert_raises_regex(ValueError, msg):
+            period_range(start='2017Q1')
+
+        with tm.assert_raises_regex(ValueError, msg):
+            period_range(end='2017Q1')
+
+        with tm.assert_raises_regex(ValueError, msg):
+            period_range(periods=5)
+
+        with tm.assert_raises_regex(ValueError, msg):
+            period_range()
```

Comment:
add inconsistent freq w.r.t. start/end

Reply:
The current behavior on 0.20.3 (and this PR as-is) is not to raise, but to convert to `freq`:

```python
In [2]: p1 = pd.Period('2017-01-01', freq='W')
   ...: p2 = pd.Period('2017-12-31', freq='W')
   ...: pd.period_range(start=p1, end=p2, freq='Q')
   ...:
Out[2]: PeriodIndex(['2017Q1', '2017Q2', '2017Q3', '2017Q4'], dtype='period[Q-DEC]', freq='Q-DEC')
```

When `freq` isn't specified, it converts to daily (default is `freq='D'`):
```python
In [5]: pd.period_range(start=p1, end=p2)
Out[5]:
PeriodIndex(['2017-01-01', '2017-01-02', '2017-01-03', '2017-01-04',
             '2017-01-05', '2017-01-06', '2017-01-07', '2017-01-08',
             '2017-01-09', '2017-01-10',
             ...
             '2017-12-22', '2017-12-23', '2017-12-24', '2017-12-25',
             '2017-12-26', '2017-12-27', '2017-12-28', '2017-12-29',
             '2017-12-30', '2017-12-31'],
            dtype='period[D]', length=365, freq='D')
```

Same behavior with different frequencies for all of `start`/`end`/`freq`:

```python
In [4]: p3 = pd.Period('2017-12-31', freq='M')
   ...: pd.period_range(start=p1, end=p3, freq='Q')
   ...:
Out[4]: PeriodIndex(['2017Q1', '2017Q2', '2017Q3', '2017Q4'], dtype='period[Q-DEC]', freq='Q-DEC')
```

Should this raise if any of `start`/`end`/`freq` are passed with different frequencies?  If `start`/`end` are passed with the same frequency, and `freq` isn't specified, should `freq` be set to match `start`/`end` instead of using the daily default?

Couldn't find a direct comparison with the other range functions to check against for consistency.  `interval_range` will raise if `freq` isn't compatible with they type of `start`/`end`, but that seems a bit different.

Reply:
I guess this is meant to be the 'new' freq, while the start/end serve as the anchor endpoints. this seems fine then. (maybe adding the above example to the docs would be helpful though).

Reply:
Okay, will add tests for this behavior.